### PR TITLE
Proceed on unresolvable hostname

### PR DESCRIPTION
**Pull request by wprzytula**
A user encountered a following problem:
> Rust Driver wasn't able to start if one of the configured nodes was not available

This was found to be caused by the hostname resolution logic, which considered failure in resolving any hostname to be a critical error.
As it makes more sense to fail only when we end up with no contact points with known resolved addresses, in that case, a new variant is returned: `NewSessionError::FailedToResolveAnyHostname`, else initialisation continues successfully.
The new behaviour is automatically tested.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.~~
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- ~~[ ] I added appropriate `Fixes:` annotations to PR description.~~

### CODE REVIEW

**Comment by mykaul** [comment]
0.8.x material?

**Comment by wprzytula** [comment]
> 0.8.x material?

This unfortunately contains an API-breaking change, so we can release it only with a new major number, i.e. 0.9.x.
WDYT @piodul ?

**Review by piodul — CHANGES_REQUESTED** [review]

File: scylla/src/transport/session.rs [hunk 1]
```diff
@@ -470,16 +471,30 @@ impl Session {
...
+        let resolve_futures = to_resolve.iter().map(|(hostname, datacenter)| async move {
+            match resolve_hostname(hostname).await {
+                Ok(address) => Ok::<_, NewSessionError>(ContactPoint {
+                    address,
+                    datacenter: datacenter.clone(),
+                }),
+                Err(e) => {
+                    warn!("Hostname resolution failed for {}: {}", hostname, &e);
+                    Err(e.into())
+                }
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/src/transport/session.rs:483]: I believe `Result` conveys the intention better. After all, `Err` points at some kind of failure, while `None` can as well signify a correct situation. I would stick to `Result`.

[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs:483]: Not sure if I agree. We are ignoring the errors, after all - we are ignoring the hostnames for which resolution failed and using only the successful ones. `filter_map` + returning `Option` is a common pattern which I think is pretty clear. Returning `Result` here only suggests that you are going to do something useful with the error later.

[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/src/transport/session.rs]: I'm convinced now. Adjusted this.

File: scylla/tests/integration/new_session.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    session
+        .query("SELECT host_id FROM system.local", &[])
+        .await
+        .unwrap();
+}
+
+#[cfg(not(scylla_cloud_tests))]
+#[tokio::test]
+async fn all_hostnames_invalid() {
+    let uri = "cassandrasuckssomu.ch:9042".to_owned();
```
Comment (piodul) [inline scylla/tests/integration/new_session.rs:28]: This one is not fixed yet.

**Review by piodul — APPROVED** [review]

**Pull request merged**